test/params_test.c

```diff
@@ -0,0 +1,549 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * https://www.openssl.org/source/license.html
+ * or in the file LICENSE in the source distribution.
+ */
+
+/*
+ * This program tests the use of OSSL_PARAM, currently in raw form.
+ */
+
+#include <string.h>
+#include <openssl/bn.h>
+#include <openssl/core.h>
+#include <openssl/params.h>
+#include "internal/nelem.h"
+#include "testutil.h"
+
+/*-
+ * PROVIDER SECTION
+ * ================
+ *
+ * Even though it's not necessarily ONLY providers doing this part,
+ * they are naturally going to be the most common providers of
+ * set_params and get_params functions.
+ */
+
+/*
+ * In real use cases, setters and getters would take an object with
+ * which the parameters are associated.  This structure is a cheap
+ * simulation.
+ */
+struct object_st {
+    /*
+     * Documented as a native integer, of the size given by sizeof(int).
+     * Assumed data type OSSL_PARAM_INTEGER
+     */
+    int p1;
+    /*
+     * Documented as a native double, of the size given by sizeof(double).
+     * Assumed data type OSSL_PARAM_REAL
+     */
+    double p2;
+    /*
+     * Documented as an arbitrarly large unsigned integer.
+     * The data size must be large enough to accomodate.
+     * Assumed data type OSSL_PARAM_UNSIGNED_INTEGER
+     */
+    BIGNUM *p3;
+    /*
+     * Documented as a C string.
+     * The data size must be large enough to accomodate.
+     * Assumed data type OSSL_PARAM_UTF8_STRING
+     */
+    char *p4;
+    size_t p4_l;
+    /*
+     * Documented as a C string.
+     * Assumed data type OSSL_PARAM_UTF8_STRING
+     */
+    char p5[256];
+    size_t p5_l;
+    /*
+     * Documented as a pointer to a constant C string.
+     * Assumed data type OSSL_PARAM_UTF8_PTR
+     */
+    const char *p6;
+    size_t p6_l;
+};
+
+#define p1_init 42                              /* The ultimate answer */
+#define p2_init 7.0                             /* Magic number */
```

Comment:
An integer for a real? Here and later. Is there a sudden shortage of non-integral real values?

Reply:
Bah...  or well, I can always use PI :wink: